[Doc Improvement][Update preview manifest to support context menu][3679844] (#11204)
MicrosoftDocs/msteams-docs: msteams-platform/resources/schema/manifest-schema-dev-preview.md

**Optional** &ndash; Object

> The `extensions` property specifies Outlook add-ins within an app manifest and simplifies the distribution and acquisition across the Microsoft 365 ecosystem. Each app supports only one extension.

Reviewer: Just noticed this. We're going to have to change this to Office in a few weeks anyway and Outlook doesn't support the context menus anyway, so let's do this now. 

```suggestion
The `extensions` property specifies Office add-ins within an app manifest and simplifies the distribution and acquisition across the Microsoft 365 ecosystem. Each app supports only one extension.
```

|`autoRunEvents`| Array | | | Defines the event-based activation extension point. |
|`alternates`| Array | | | Specifies the relationship to alternate existing Microsoft 365 solutions. It's used to hide or prioritize add-ins from the same publisher with overlapping functionality. |
|`audienceClaimUrl`| String | 2048 characters | | Specifies the URL for your extension and is used to validate Exchange user identity tokens. For more information, see [inside the Exchange identity token](/office/dev/add-ins/outlook/inside-the-identity-token)|

> |`contextMenus`| Array | | | Specifies the context menus for your extension. A context menu is a shortcut menu that appears when you right-click in the Office UI. |

Reviewer 2: Is it 'right-click' within the Office UI?

Author: Not sure about using "within" but updated instruction according to [MS writing guidance](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/r/right-click)

Author: Opting to keep "in" per our editor and [guidance](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/term-collections/mouse-mouse-interaction-terms)

> |`contextMenus`| Array | | | Specifies the context menus for your extension. A context menu is a shortcut menu that appears when you select and hold (or right-click) in the Office UI. |

Reviewer 3: Is "Office UI" a specific concept, or is it meant here to stand for the UI of the given WXPO host app? If the latter, how about something like `UI of the host application`? /Cc @v-kushals

Reviewer: The right click will always be in the open Office document. They won't be right clicking the Office ribbon or anything else in the Office UI, so I would recommend replacing "in the Office UI" with "in the open Office document".

> For more information, see [Office Add-ins manifest for Microsoft 365](/office/dev/add-ins/develop/unified-manifest-overview).

Reviewer 2: Hi @ElizabethSamuel-MSFT ,
Observed that the feature document referred here does not contain 'contextMenus' under the section "extensions" property and in the sample unified manifest. Please do the needful.

Reviewer: @v-ramyas5 
The unified-manifest-overview article is intended to be an overview, not a replacement for the reference documentation of the “extensions” property. I will update it to make it clear that the table shows only certain selected and representative child properties, not all of them.

We should not update the sample manifest. It is not possible to have a sample with every possible property anyway because some are mutually exclusive. Note that the sample manifests in the overview for the add-in only manifest also don’t have every possible element.

Reviewer 2: Hey @Rick-Kirkham,
Sorry for the delayed response. I understand that the doc may not have all the fields. It is nice-to-add context menu to help new developers understand the feature better. Just a small suggestion from our side :)

### extensions.requirements

The `extensions.requirements` object specifies the scopes, form factors, and Office JavaScript Library requirement sets that must be supported on the Office client in order for the add-in to be installed. Requirements are also supported on the "ribbon", "runtime", "alternates", "autoRunEvents", and "contextMenus" child properties to selectively filter out some features of the add-in. For more information, see [Specify Office Add-in requirements in the unified manifest for Microsoft 365](/office/dev/add-ins/develop/requirements-property-unified-manifest).

| Name | Type | Maximum size | Required | Description |
|---|---|---|---|---|
|`requirements.capabilities`| Array | | | Identifies the requirement sets. <br>Options: `name` (required), `minVersion`, `maxVersion`|

> |`requirements.capabilities.name`| String | | ✔️ | Identifies the name of the requirement set. |

Author: Further to Ricky's comment re: contextMenus.requirements.scopes, will we need to update the requirements.scopes description in this section? It currently says, "For example, `mail` (Outlook). <br>Supported value: `mail`". Should it include the WXP values as well?

Reviewer: > Further to Ricky's comment re: contextMenus.requirements.scopes, will we need to update the requirements.scopes description in this section (currently line 972)? It currently says, "For example, `mail` (Outlook). Supported value: `mail`". Should it include the WXP values as well?

Yes. I've made that update in another PR. These differences will appear as merge conflicts, but I think it is easier to wait from those conflicts to arise and fix them, than to try and write every PR so that it includes all the changes from every other PR.

|`alternateIcons.icon.url`| String | 2048 characters | ✔️ | Specifies the full, absolute URL of the image file that is used to represent the add-in. Icon image must be 64 x 64 pixels and use one of the following file formats: GIF, JPG, PNG, EXIF, BMP, TIFF.|
|`alternateIcons.highResolutionIcon`| Object | | ✔️ | Specifies properties of the image file used to represent the add-in on high DPI screens. |
|`alternateIcons.highResolutionIcon.size`| Number enum | | ✔️ | This property is reserved for future use. The value must be 64 (not 128). |
|`alternateIcons.highResolutionIcon.url`| String | 2048 characters | ✔️ | Specifies the full, absolute URL of the image file that is used to represent the add-in on high DPI screens. Icon image must be 128 x 128 pixels and use one of the following file formats: GIF, JPG, PNG, EXIF, BMP, TIFF.|

### extensions.contextMenus

A context menu is a shortcut menu that appears when you right-click (or select and hold) in an open Office document.

|Name | Type | Maximum size | Required | Description |
|---|---|---|---|---|
|`menus`| Array | | ✔️ | Configures the context menus. |

> |`menus.entryPoint`| String enum | | ✔️ | Use `text` or `cell` here for Office context menu. Use `text` if the context menu should open when a user right-clicks (or selects and holds) on the selected text. Use `cell` if the context menu should open when the user right-clicks (or selects and holds) on a cell in an Excel spreadsheet. |

Reviewer: @v-ramyas5 and @ElizabethSamuel-MSFT , I assume the question is what happens if the user right clicks a *table cell* in Word or PPT. Since that entryPoint is not supported in Word or PPT, the context menus would not appear.

> |`menus.controls`| Array | | ✔️ | The same as controls in [extensions.ribbons.tabs.groups.controls](#extensionsribbons). The control type should be `menu`. |

Reviewer: This traces back to a mistake I made in XML reference a while back. It's corrected there in a PR this week.

```suggestion
|`menus.controls`| Array | | ✔️ | The same as controls in [extensions.ribbons.tabs.groups.controls](#extensionsribbons). The control type should be `button`. |
```

> |`menus.controls`| Array | | ✔️ | The same as controls in [extensions.ribbons.tabs.groups.controls](#extensionsribbons). The control type should be `button`. |

Reviewer 2: Hey @ElizabethSamuel-MSFT,
First of all, apologies for the recurring reviews as we couldn't finish it in one-go due to other priorities. I will make arrangements to review and merge this PR asap. 

In this line: 1132, though I really like the idea of linking to a similar property (I have also proposed it once), we finally decided to explicitly list all the properties with the same data so the developers can cross-check and make sure the properties, default values, max size are specific and tailored to the specific feature they are defining. Request you to list the sub-properties of `menus.controls` for additional clarity for developers.

Author: @v-ramyas5 I've added all the extensions.ribbons.tabs.groups.controls entries here, updating them to menus.controls.*. However, I'll need some help to understand which should be removed/updated. Can you and @Rick-Kirkham help with that?

Reviewer: @ElizabethSamuel-MSFT , I did this in the form of a very long suggestion comment.

|`alternateIcons.highResolutionIcon.size`| Number enum | | ✔️ | This property is reserved for future use. The value must be 64 (not 128). |
|`alternateIcons.highResolutionIcon.url`| String | 2048 characters | ✔️ | Specifies the full, absolute URL of the image file that is used to represent the add-in on high DPI screens. Icon image must be 128 x 128 pixels and use one of the following file formats: GIF, JPG, PNG, EXIF, BMP, TIFF.|

### extensions.contextMenus

> A context menu is a shortcut menu that appears when you select and hold (or right-click) in the Office UI.

Reviewer 3: Same for this instance. Just want to disambiguate from "Microsoft 365 (Office)" app (Harmony)

|Name | Type | Maximum size | Required | Description |
|---|---|---|---|---|
|`menus`| Array | | ✔️ | Configures the context menus. |
|`menus.entryPoint`| String enum | | ✔️ | Use `text` or `cell` here for Office context menu. Use `text` if the context menu should open when a user right-clicks on the selected text. Use `cell` if the context menu should open when the user right-clicks on a cell in an Excel spreadsheet. |

> |`menus.controls`| Array | | ✔️ | The same as controls in ribbons. The control type should be `menu`. |

Reviewer: ```suggestion
|`menus.controls`| Array | | ✔️ | The same as controls in [extensions.ribbons.tabs.groups.controls](#extensionsribbons). The control type should be `menu`. |
```

|`requirements`| Object | | | Specifies the scopes, formFactors, and Office JavaScript Library requirement sets that must be supported on the Office client in order for the context menus to appear. For more information, see [Specify Office Add-in requirements in the unified manifest for Microsoft 365](/office/dev/add-ins/develop/requirements-property-unified-manifest). |
|`requirements.capabilities`| Array | | | Identifies the requirement sets. <br>Options: `name` (required), `minVersion`, `maxVersion`|
|`requirements.capabilities.name`| String | | ✔️ | Identifies the name of the requirement set. |
|`requirements.capabilities.minVersion`| String | | | Identifies the minimum version for the requirement set. |
|`requirements.capabilities.maxVersion`| String | | | Identifies the maximum version for the requirement set. |

> |`requirements.scopes`| Array of enums | 1 | | Identifies the scopes in which the add-in can run and defines the Microsoft 365 applications in which the extension can run. For example, `mail` (Outlook). <br>Supported value: `mail` |

Reviewer: Sorry. Just noticed another. 
```suggestion
|`requirements.scopes`| Array of enums | 1 | | Identifies the scopes in which the add-in can run and defines the Microsoft 365 applications in which the extension can run. For example, `workbook` (Excel). <br>Supported values: `workbook`, `presentation`, `document`. |
```

|Name| Type| Maximum size | Required | Description|
|---|---|---|---|---|

> |`events`| Array |20 | ✔️ | Configures the event that cause actions in an Outlook add-in to run automatically. For example, see [use smart alerts and the `OnMessageSend` and `OnAppointmentSend` events in your Outlook add-ins](/office/dev/add-ins/outlook/smart-alerts-onmessagesend-walkthrough?tabs=jsonmanifest).|

Reviewer 3: Looks like "Add-ins" should be capitalized?

Author: No, it doesn't. We only capitalize Office Add-ins or Office Add-in. The others are lowercase, e.g., Outlook add-ins, Excel add-ins.

Reviewer 3: Ah, ok! Good to know.

|`alternateIcons.highResolutionIcon.url`| String | 2048 characters | ✔️ | Specifies the full, absolute URL of the image file that is used to represent the add-in on high DPI screens. Icon image must be 128 x 128 pixels and use one of the following file formats: GIF, JPG, PNG, EXIF, BMP, TIFF.|

### extensions.contextMenus

A context menu is a shortcut menu that appears when you right-click (or select and hold) in an open Office document. Currently valid only for Word, Excel, and PowerPoint.

|Name | Type | Maximum size | Required | Description |
|---|---|---|---|---|
|`menus`| Array | | ✔️ | Configures the context menus. |
|`menus.entryPoint`| String enum | | ✔️ | Defines the method for opening the context menu. Use `text` if the context menu should open when a user right-clicks (or selects and holds) on the selected text. Use `cell` if the context menu should open when the user right-clicks (or selects and holds) on a cell in an Excel spreadsheet. |
|`menus.controls`| Array | | ✔️ | Configures the menu buttons. |

> |`tabs.groups.controls.id`| String | 64 characters | ✔️ | Specifies the ID for the control within the app. It must be different from any built-in control ID in the Microsoft 365 application and any other custom control. |
> |`tabs.groups.controls.items`| Array | | | Configures the items for a menu control. |
> |`tabs.groups.controls.items.id`| String | | ✔️ | Specifies the ID for the items within the app. |
> |`tabs.groups.controls.items.type`| String enum | | ✔️ | Defines the control items type. <br>Supported values: `button`, `menu`|
> |`tabs.groups.controls.items.label`| String | 64 characters | ✔️ | Specifies the text displayed for the items. |
> |`tabs.groups.controls.items.icons`| Array | | | Configures the icons for the custom item.|
> |`tabs.groups.controls.items.icons.size`| Number | | ✔️ | Specifies the size of the icon in pixels, enumerated as `16`,`20`,`24`,`32`,`40`,`48`,`64`,`80`. <br>Required image sizes: `16`, `32`, `80`. |
> |`tabs.groups.controls.items.icons.url`| String | 2048 characters | ✔️ | Specifies the absolute URL of the icon.|
> |`tabs.groups.controls.items.supertip`| | | ✔️ | Configures a supertip for the custom item. A supertip is a UI feature that displays a brief box of help information about a control when the cursor hovers over it. The box may contain multiple lines of text. |
> |`tabs.groups.controls.items.supertip.title`| String | 64 characters | ✔️ | Specifies the title text of the supertip.|
> |`tabs.groups.controls.items.supertip.description`| String | 250 characters | ✔️ | Specifies the description of the supertip.|
> |`tabs.groups.controls.items.actionId`| String | 64 characters | ✔️ | Specifies the ID of the action that is taken when a user selects the control or menu item. The `actionId` must match with `runtime.actions.id`. |
> |`tabs.groups.controls.items.enabled`| Boolean | | | Indicates whether the control is initially enabled. <br>Default value: `true`|
> |`tabs.groups.controls.items.overriddenByRibbonApi`| Boolean | | | Specifies whether a group, button, menu, or menu item hidden on application and platform combinations, which support the API ([Office.ribbon.requestCreateControls](/javascript/api/office/office.ribbon#office-office-ribbon-requestcreatecontrols-member(1))) that installs custom contextual tabs on the ribbon. <br>Default value: `false`|
> |`tabs.groups.controls.type`| String | | ✔️ | Defines the control type. <br>Supported values: `button`|
> |`tabs.groups.controls.builtinControlId`| String | 64 characters | ✔️ | Specifies the ID of an existing Microsoft 365 control. For more information, see [find the IDs of controls and control groups](/office/dev/add-ins/design/built-in-button-integration#find-the-ids-of-controls-and-control-groups).|
> |`tabs.groups.controls.label`| String | 64 characters | ✔️ | Specifies the text displayed for the control. Despite the maximum length of 64 characters, to correctly align the tab in the ribbon, we recommend you limit the label to 16 characters.|
> |`tabs.groups.controls.icons`| Array | | ✔️ | Defines the icons for the control. There must be at least three child objects; one each with `size` properties of `16`, `32`, and `80` pixels. |
> |`tabs.groups.controls.icons.size`| Number | | ✔️ | Specifies the size of the icon in pixels, enumerated as `16`,`20`,`24`,`32`,`40`,`48`,`64`,`80`. <br> Required image size: `16`, `32`, `80`|
> |`tabs.groups.controls.icons.url`| String | 2048 characters | | Specifies the absolute URL to the icon.|
> |`tabs.groups.controls.supertip`| Object | | ✔️ | Configures a supertip for the control. |
> |`tabs.groups.controls.supertip.title`| String | 64 characters | ✔️ |Specifies the title text of the supertip.|
> |`tabs.groups.controls.supertip.description`| String | 128 characters | ✔️ | Specifies the description of the supertip.|
> |`tabs.groups.controls.actionId`| String | 64 characters | | Required if the control type is `button`. Don't use if the control type is `menu`. Specifies the ID of the action that is taken when a user selects the control. The `actionId` must match the `runtime.actions.id` property of an action in the `runtimes` object.|
> |`tabs.groups.controls.enabled`| Boolean | | | Indicates whether the control is initially enabled. <br>Default value: `true`|
> |`tabs.groups.controls.overriddenByRibbonApi`| Boolean | | | Specifies whether a group, button, menu, or menu item is hidden on application and platform combinations which support the API ([Office.ribbon.requestCreateControls](/javascript/api/office/office.ribbon#office-office-ribbon-requestcreatecontrols-member(1))) that installs custom contextual tabs on the ribbon. <br>Default value: `false`|

Reviewer: ```suggestion
|`menus.controls.id`| String | 64 characters | ✔️ | Specifies the ID for the menu control within the app. It must be different from any built-in control ID in the Microsoft 365 application and any other custom control. |
|`menus.controls.items`| Array | | | Configures the items for the menu control. |
|`menus.controls.items.id`| String | | ✔️ | Specifies the ID for a menu item. |
|`menus.controls.items.type`| String enum | | ✔️ | Defines the item's control type. <br>Supported values: `button`|
|`menus.controls.items.label`| String | 64 characters | ✔️ | Specifies the text displayed for the item. |
|`menus.controls.items.icons`| Array | | | Configures the icons for the custom item.|
|`menus.controls.items.icons.size`| Number | | ✔️ | Specifies the size of the icon in pixels, enumerated as `16`,`20`,`24`,`32`,`40`,`48`,`64`,`80`. <br>Required image sizes: `16`, `32`, `80`. |
|`menus.controls.items.icons.url`| String | 2048 characters | ✔️ | Specifies the absolute URL of the icon.|
|`menus.controls.items.supertip`| | | ✔️ | Configures a supertip for the custom item. A supertip is a UI feature that displays a brief box of help information about a control when the cursor hovers over it. The box may contain multiple lines of text. |
|`menus.controls.items.supertip.title`| String | 64 characters | ✔️ | Specifies the title text of the supertip.|
|`menus.controls.items.supertip.description`| String | 250 characters | ✔️ | Specifies the description of the supertip.|
|`menus.controls.items.actionId`| String | 64 characters | ✔️ | Specifies the ID of the action that is taken when a user selects the menu item. The `actionId` must match with `runtime.actions.id`. |
|`menus.controls.items.enabled`| Boolean | | | Indicates whether the menu item is initially enabled. <br>Default value: `true`.|
|`menus.controls.items.overriddenByRibbonApi`| Boolean | | | Specifies whether the menu item is hidden on application and platform combinations which support the API ([Office.ribbon.requestCreateControls](/javascript/api/office/office.ribbon#office-office-ribbon-requestcreatecontrols-member(1))) that installs custom contextual tabs on the ribbon. <br>Default value: `false`|
|`menus.controls.type`| String | | ✔️ | Defines the control type. <br>Supported values: `menu`.|
|`menus.controls.builtInControlId`| String | 64 characters | ✔️ | Specifies the ID of an existing Microsoft 365 control. For more information, see [find the IDs of controls and control groups](/office/dev/add-ins/design/built-in-button-integration#find-the-ids-of-controls-and-control-groups). If this property is present, no other "menus.*" child properties may be used because a built-in Office control is not customizable by an add-in.|
|`menus.controls.label`| String | 64 characters | ✔️ | Specifies the text displayed for the menu control.|
|`menus.controls.icons`| Array | | ✔️ | Defines the icons for the menu control. There must be at least three child objects; one each with `size` properties of `16`, `32`, and `80` pixels. |
|`menus.controls.icons.size`| Number | | ✔️ | Specifies the size of the icon in pixels, enumerated as `16`,`20`,`24`,`32`,`40`,`48`,`64`,`80`. <br> Required image size: `16`, `32`, `80`|
|`menus.controls.icons.url`| String | 2048 characters | | Specifies the absolute URL to the icon.|
|`menus.controls.supertip`| Object | | ✔️ | Configures a supertip for the menu control. |
|`menus.controls.supertip.title`| String | 64 characters | ✔️ |Specifies the title text of the supertip.|
|`menus.controls.supertip.description`| String | 128 characters | ✔️ | Specifies the description of the supertip.|
|`menus.controls.actionId`| String | 64 characters | | Don't use. The action of the menu control is always the same: it displays the list of menu items.|
|`menus.controls.enabled`| Boolean | | | Indicates whether the control is initially enabled. <br>Default value: `true`.|
|`menus.controls.overriddenByRibbonApi`| Boolean | | | Specifies whether the menu is hidden on application and platform combinations which support the API ([Office.ribbon.requestCreateControls](/javascript/api/office/office.ribbon#office-office-ribbon-requestcreatecontrols-member(1))) that installs custom contextual tabs on the ribbon. <br>Default value: `false`.|
```
